### PR TITLE
fix: box_syntax/pattern

### LINKED ISSUE #1412: box_syntax/pattern not supported
```rust
let x: Box<i32> = box 1i32;
let box y = x;
```
cc https://github.com/rust-lang/rust/issues/49733

### PR DESCRIPTION
**Pull request by csmoe**
Closes #1412 
r? @matklad

### CODE REVIEW

File: crates/ra_parser/src/grammar/patterns.rs [hunk 1]
```diff
@@ -260,11 +261,13 @@ fn pat_list(p: &mut Parser, ket: SyntaxKind) {
 //     let ref mut d = ();
 //     let e @ _ = ();
 //     let ref mut f @ g @ _ = ();
+//     let box i = Box::new(1i32);
 // }
 fn bind_pat(p: &mut Parser, with_at: bool) -> CompletedMarker {
     let m = p.start();
     p.eat(T![ref]);
     p.eat(T![mut]);
+    p.eat(T![box]);
```
Comment (bjorn3) [inline crates/ra_parser/src/grammar/patterns.rs]: The `box` must be eated before the rest.

```
let box ref mut a = Box::new(());
```

is valid rust. Proof: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d2775a785a658a922cdc958e76f8dfa3

Comment (csmoe) [reply crates/ra_parser/src/grammar/patterns.rs]: thanks ❤️

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by Robbepop** [comment]
Does this PR also introduce box patterns in the way of matching them by box patterns?
I fail to see a test for this.
Something like this:
https://github.com/Robbepop/stevia/blob/master/simplifier/src/simplifications/bool_symbolic_solver.rs#L52

**Comment by matklad** [comment]
I think
https://github.com/rust-analyzer/rust-analyzer/pull/1414/files#diff-2363d5bd87246fdd9203ebd8cf0fc230R264
covers this?

**Comment by Robbepop** [comment]
Ah got it, didn't notice it was covered by that! Thanks. :)

File: crates/ra_parser/src/grammar/expressions/atom.rs [hunk 2]
```diff
@@ -507,3 +508,17 @@ fn try_block_expr(p: &mut Parser, m: Option<Marker>) -> CompletedMarker {
     block(p);
     m.complete(p, TRY_EXPR)
 }
+
+// test box_expr
+// fn foo() {
+//     let x = box 1i32;
+// }
+fn box_expr(p: &mut Parser, m: Option<Marker>) -> CompletedMarker {
```
Comment (csmoe) [inline crates/ra_parser/src/grammar/expressions/atom.rs]: @matklad oops...did I miss something in this method? box syntax inside "tuple"s cannot be parsed.
```rust
struct Foo(Box<i32>);
let a = Foo(box 1i32);
let b = (box 1i32, box 2i32);
```

**Comment by matklad** [comment]
Hm, I guess, we should add box to expr_first ?

**Pull request merged**